packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -560,8 +556,23 @@ class OffsetLayer extends ContainerLayer {
   }
 
   @override
-  void addToScene(ui.SceneBuilder builder, Offset layerOffset) {
-    addChildrenToScene(builder, offset + layerOffset);
+  void applyTransform(Layer child, Matrix4 transform) {
+    assert(child != null);
+    assert(transform != null);
+    transform.multiply(Matrix4.translationValues(offset.dx, offset.dy, 0.0));
+  }
+
+  @override
+  void addToScene(ui.SceneBuilder builder) {
+    // Skia has a fast path for concatenating scale/translation only matrices.
+    // Hence this operation should be fast. For retained rendering, we don't
+    // want to push the offset down to each leaf node. Otherwise, changing an
+    // offset layer on the very high level could cascade the change to too many
+    // leaves.
+    final Matrix4 matrix = new Matrix4.translationValues(offset.dx, offset.dy, 0.0);
+    builder.pushTransform(matrix.storage);
```

Comment:
I think I'd prefer a dedicated `pushOffset` method on `SceneBuilder` that matches the API of `OffsetLayer`:

```dart
@override
void addToScene(ui.SceneBuilder builder) {
  builder.pushOffset(offset.dx, offset.dy);
}
```

The `dx` and `dy` can be easily (and more efficiently) converted to a transform matrix by the engine. Knowing the difference between a transform and an offset I can use it to collapse a series of offsets into a single offset as an optimization that reduces the total number of layers.

Reply:
Sounds good to me. https://github.com/flutter/engine/pull/6349 is uploaded to handle this. Once that landed, I can change this (either in this PR, or in the next PR).